### PR TITLE
Fix printer column generation for struct fields that are serialized as strings

### PR DESCRIPTION
**Pull request by brandond**
Kubernetes types like metav1.Time and resource.Quantity are serialized as strings, but the column tag extraction did not handle this properly. Fix this by using the OpenAPISchemaType helper function to get the type.

Ref: https://github.com/kubernetes/kube-openapi/blob/master/pkg/generators/README.md#custom-openapi-type-definitions

### CODE REVIEW

**Review by KevinJoiner — CHANGES_REQUESTED** [review]
These changes make sense. We have been requiring Unit tests for all affected code in our PRs

> but the column tag extraction did not handle this properly.

Can you provide an example of what the broken behavior looked like.

File: pkg/crd/init.go
Comment (brandond) [reply pkg/crd/init.go:154]: I looked through the core Kubernetes codebase and couldn't find anything that returned anything other than just `["string"`] - even `intOrString` just returns `["string"]`

* https://github.com/kubernetes/apimachinery/blob/master/pkg/util/intstr/intstr.go#L133
* https://github.com/kubernetes/apimachinery/blob/master/pkg/api/resource/quantity.go#L405
* https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/time.go#L168
* https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/micro_time.go#L167
* https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/duration.go#L61

File: pkg/crd/init.go
Comment (brandond) [reply pkg/crd/init.go:118]: No, without a json path the column doesn't do anything, so the `column` tag effectively required the jsonpath to be set manually whenever used. This, combined with fixing the path generation code up above, allows for generation of columns with default name, type, and path by just adding a bare `column:` tag.

File: pkg/crd/init.go
Comment (brandond) [reply pkg/crd/init.go:153]: I was unable to find any core Kubernetes types that use that interface.

File: pkg/crd/init.go
Comment (brandond) [reply pkg/crd/init.go:152]: Yeah, but `string` is the only thing that I can find anything using a the moment. Maybe cross that bridge when we come to it?

File: pkg/crd/init.go
Comment (rmweir) [inline pkg/crd/init.go:184]: Was this just incorrect before? I the topmost call would already pass "." as the path.

**Review by rmweir — CHANGES_REQUESTED** [review]
Once @KevinJoiner feedback has been addressed, LGTM. Had one question about a fix but the actual change looks good.

**Comment by brandond** [comment]
> Can you provide an example of what the broken behavior looked like.

It would just skip the column completely when using `WithColumnsFromStruct()`

File: pkg/crd/init.go
Comment (brandond) [reply pkg/crd/init.go:184]: Yes, this was wrong before. The root path is . so it would generate things like ..spec instead of .spec. as the parent path for nested fields.

File: pkg/crd/init.go
Comment (brandond) [reply pkg/crd/init.go:117]: Added!

**Comment by brandond** [comment]
I added format support, and a battery of tests. PTAL!

File: pkg/crd/init.go
Comment (brandond) [reply pkg/crd/init.go:117]: It turns out the format returned here is intended to be used by the full schema, but the printer columns only support the core data types: https://github.com/OAI/OpenAPI-Specification/blob/main/versions/2.0.md#data-types

File: pkg/crd/init.go
Comment (KevinJoiner) [reply pkg/crd/init.go:187]: Yeah, I also noticed that the path variable just wasn't used before 😆

File: pkg/crd/init.go
Comment (KevinJoiner) [reply pkg/crd/init.go:157]: Okay this should be fine, I would still rather only accept a slice of len == 1 since structs outside of ones defined in upstream can also hit this, but since this isn't a library used by people outside of Rancher this should be fine.

File: pkg/crd/init.go
Comment (KevinJoiner) [reply pkg/crd/init.go:155]: Cool, I am okay with this since we don't have to support 1:1 compatibility with kube-api

**Review by KevinJoiner — CHANGES_REQUESTED** [review]
This looks good once we figure out the format stuff. Thanks for adding the unit test they have good coverage and are easy to extend if someone else needs to add additional tets.

File: pkg/crd/init.go
Comment (KevinJoiner) [reply pkg/crd/init.go:117]: I think the mapping should work I was curious what happens if we add an invalid format and it looks like api-server validation blocks it 
```
spec.additionalPrinterColumns[1].format: Invalid value: "int-or-string": must be one of byte,date,date-time,double,float,int32,int64,password
```

**Review by rmweir — APPROVED** [review]
LGTM, once @KevinJoiner has approved.

File: pkg/crd/init.go
Comment (brandond) [reply pkg/crd/init.go]: Yeah, I noticed that as well when actually trying to load one of the generated CRDs into a cluster.

I've added handling of supported types and formats, and tests now call into upstream's validation code to ensure that the columns will actually be accepted by the apiserver.

**Review by KevinJoiner — APPROVED** [review]
Looks good!

**Pull request merged**
